cache-prefetch.go

```diff
@@ -0,0 +1,183 @@
+	mu       sync.Mutex
+	metrics  *CachePrefetchMetrics
+}
+
+
+type CachePrefetchMetrics struct {
+	// Cache hit count.
+	domainEntries map[string]CachePrefetchEntry
+}
+type CachePrefetchEntry struct {
+	// request hit count.
+	hit 			  expvar.Int
+
+	// 0 for no prefetching job
+	// 1 for prefetching job is active
+	// 2 for stopped by error
+	prefetchingStatus expvar.Int
+	// store the time to live
+	ttl				  expvar.Int
+	// fetching error count for discarding error prone fetches
+	errorCount 		  expvar.Int
+}
+
+var _ Resolver = &CachePrefetch{}
+
+type CachePrefetchOptions struct {
+	//// Time of cache record ttl polling for record prefetch
+	CacheTTLPollingCheckInterval time.Duration
+	//// Min record time remaining check for expire
+	MinRecordTimeRemainingPercent uint64
+
+	// Number of hits a record gets before prefetch on a record is started
+	RecordQueryHitsMin int64
+	// Max number of responses to check in the cache. Defaults to 0 which means no limit. If
+	// the limit is reached, the least-recently used entry is removed from the cache.
+	//TODO
+	//RecordCheckCapacity uint32
+	// Cache to check records from
+	//CacheResolver Cache
+	// Allows control over the order of answer RRs in cached responses. Default is to keep
+	// the order if nil.
+	// TODO
+	//ShuffleAnswerFunc AnswerShuffleFunc
+}
+
+func NewCachePrefetch(id string, resolver Resolver, opt CachePrefetchOptions) *CachePrefetch {
+	c := &CachePrefetch{
+		CachePrefetchOptions: opt,
+		id:                   id,
+		resolver:             resolver,
+		metrics: &CachePrefetchMetrics{
+			domainEntries: map[string]CachePrefetchEntry{},
+		},
+	}
+	if c.CacheTTLPollingCheckInterval == 0 {
+		c.CacheTTLPollingCheckInterval = time.Minute
+	}
+	if c.RecordQueryHitsMin == 0 || c.RecordQueryHitsMin == 1 || c.RecordQueryHitsMin == -1 {
+		// Set to hit after one record hit
+		// fetch opportunistically
+		c.RecordQueryHitsMin = 1
+	}
+	return c
+}
+
+func (r *CachePrefetch) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
+	if len(q.Question) < 1 {
+		return nil, errors.New("no question in query")
+	}
+	// While multiple questions in one DNS message is part of the standard,
+	// it's not actually supported by servers. If we do get one of those,
+	// just pass it through and bypass caching.
+	if len(q.Question) > 1 {
+		return r.resolver.Resolve(q, ci)
+	}
+
+	go r.requestAddPrefetchJob(q, ci)
+
+	// Get a response from upstream
+	a, err := r.resolver.Resolve(q.Copy(), ci)
+	if err != nil || a == nil {
+		return nil, err
+	}
+
+	// Put the upstream response into the cache and return it. Need to store
+	// a copy since other elements might modify the response, like the replacer.
+	return a, nil
+}
+
+func (r *CachePrefetch) String() string {
+	return r.id
+}
+func (r *CachePrefetch) startCachePrefetch(q *dns.Msg, ci ClientInfo) {
+	var maxNumberOfErrorsBeforeDiscardingPrefetchJob = int64(5)
+	var qname = q.Question[0].Name
+
+	for {
+		var domainEntry = r.metrics.domainEntries[qname]
```

Comment:
I bet we need to include the query type in that key, not just the qname (otherwise it's going to treat queries for MX and A records the same). And possibly even the values from the opt records, take a look at how the cache generates a key.